File: dataforge-control/src/main/java/hep/dataforge/control/ports/Port.java
/* 
 * Copyright 2015 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package hep.dataforge.control.ports;

import hep.dataforge.exceptions.PortException;
import hep.dataforge.exceptions.PortLockException;
import hep.dataforge.meta.Meta;
import hep.dataforge.meta.Metoid;
import hep.dataforge.utils.MetaHolder;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Predicate;

/**
 * The universal asynchronous port handler
 *
 * @author Alexander Nozik
 */
public abstract class Port extends MetaHolder implements AutoCloseable, Metoid {
    /**
     * The definition of default phrase condition
     *
     * @return
     */
    private static Predicate<String> DEFAULT_PHRASE_CONDITION = (String str) -> str.endsWith("\n");

    private final ReentrantLock portLock = new ReentrantLock(true);

    private PortController controller;

    /**
     * The default end phrase condition
     */
    private Predicate<String> phraseCondition = DEFAULT_PHRASE_CONDITION;

    private ExecutorService executor = Executors.newSingleThreadExecutor(r -> {
        Thread res = new Thread(r);
        res.setName("port::" + Port.this.toString());
        res.setPriority(Thread.MAX_PRIORITY);
        return res;
    });

    protected Port(Meta meta) {
        super(meta);
    }

    protected Logger getLogger() {
        return LoggerFactory.getLogger(meta().getString("logger", getClass().getName()));
    }

    public void setPhraseCondition(Predicate<String> condition) {
        this.phraseCondition = condition;
    }

    public void setDelimiter(String delimiter) {
        phraseCondition = (String str) -> str.endsWith(delimiter);
    }

    public abstract void open() throws PortException;

    public abstract boolean isOpen();

    protected void run(Runnable r) {
        executor.submit(r);
    }

    /**
     * Emergency hold break.
     */
    public synchronized void breakHold() {
//        controller = null;
        run(portLock::unlock);
    }

    /**
     * An unique ID for this port
     *
     * @return
     */
    public abstract String toString();

    /**
     * Acquire lock on this instance of port handler with given controller
     * object. Only the same controller can release the hold.
     *
     * @param controller
     * @throws hep.dataforge.exceptions.PortException
     */
    public void holdBy(PortController controller) throws PortException {
        assert controller != null;
        if (!isOpen()) {
            open();
        }

        run(() -> {
            try {
                portLock.lockInterruptibly();
            } catch (InterruptedException ex) {
                getLogger().error("Lock on port {} is broken", toString());
                throw new RuntimeException(ex);
            }
        });
        getLogger().debug("Locked by {}", controller);
        this.controller = controller;
    }

    /**
     * The condition that should be satisfied to complete the incoming message
     *
     * @param str
     * @return
     */
    protected boolean isPhrase(String str) {
        return phraseCondition.test(str);
    }

    /**
     * This method accepts complete phrase and sends it to current controller
     *
     * @param phrase
     */
    protected synchronized void receivePhrase(String phrase) {
//        lastResponse.lazySet(phrase);
        notify();
        getLogger().trace("RECEIVE: " + phrase);
        if (controller != null) {
            controller.acceptPhrase(phrase);
        }
    }

    /**
     * send the message to the port
     *
     * @param message
     * @throws hep.dataforge.exceptions.PortException
     */
    protected abstract void send(String message) throws PortException;

    /**
     * Send the message if the controller is correct
     *
     * @param controller
     * @param message
     * @throws PortException
     */
    public final void send(@Nullable PortController controller, String message) throws PortException {
        if (this.controller == null || controller == this.controller) {
            send(message);
        } else {
            throw new PortException("Port locked by another controller");
        }
    }

    /**
     * Release hold of this portHandler from given controller.
     *
     * @param controller
     * @throws PortLockException in case given holder is not the one that holds
     *                           handler
     */
    public synchronized void releaseBy(PortController controller) throws PortLockException {
        if (isLocked()) {
            assert controller != null;
            if (controller.equals(this.controller)) {
                this.controller = null;
                run(() -> {
                    portLock.unlock();
                    getLogger().debug("Unlocked by {}", controller);
                });
            } else {
                throw new PortLockException("Can't unlock hold with wrong holder");
            }
        } else {
            getLogger().warn("Attempting to release unlocked port");
        }
    }

    public boolean isLocked() {
        return this.controller != null;
    }


    @Override
    public void close() throws Exception {
        executor.shutdown();
    }

    /**
     * The controller which is currently working with this handler. One
     * controller can simultaneously hold many handlers, but handler could be
     * held by only one controller.
     */
    public interface PortController {

        void acceptPhrase(String message);

        default void portError(String errorMessage, Throwable error) {
            //do nothing
        }
    }
}

File: dataforge-control/src/main/kotlin/hep/dataforge/control/ports/Port.kt
/*
 * Copyright  2017 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hep.dataforge.control.ports

import hep.dataforge.Named
import hep.dataforge.exceptions.PortException
import hep.dataforge.exceptions.PortLockException
import hep.dataforge.meta.Meta
import hep.dataforge.meta.MetaHolder
import hep.dataforge.meta.MetaID
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.time.Duration
import java.util.concurrent.Executors
import java.util.concurrent.locks.ReentrantLock


/**
 * The controller which is currently working with this handler. One
 * controller can simultaneously hold many handlers, but handler could be
 * held by only one controller.
 */
interface PortController {

    fun accept(byte: Byte)

    fun accept(bytes: ByteArray)

    fun error(errorMessage: String, error: Throwable) {
        //do nothing
    }
}


/**
 * The universal asynchronous port handler
 *
 * @author Alexander Nozik
 */
abstract class Port(meta: Meta) : MetaHolder(meta), AutoCloseable, MetaID, Named {

    private val portLock = ReentrantLock(true)

    private var controller: PortController? = null

    protected val executor = Executors.newSingleThreadExecutor { r ->
        val res = Thread(r)
        res.name = "port::$name"
        res.priority = Thread.MAX_PRIORITY
        res
    }

    protected val logger: Logger by lazy { LoggerFactory.getLogger(meta.getString("logger", "port.$name")) }

    abstract val isOpen: Boolean

    private val isLocked: Boolean
        get() = this.portLock.isLocked
//
//    fun setPhraseCondition(condition: (String) -> Boolean) {
//        this.phraseCondition = condition
//    }
//
//    fun setDelimiter(delimiter: String) {
//        phraseCondition = { str: String -> str.endsWith(delimiter) }
//    }

    @Throws(PortException::class)
    abstract fun open()

    /**
     * Run something on port thread
     * @param r
     */
    protected fun execute(r: () -> Unit) {
        executor.submit(r)
    }

    /**
     * Emergency hold break.
     */
    @Synchronized
    fun breakHold() {
        if (isLocked) {
            logger.warn("Breaking hold on port $name")
            execute { portLock.unlock() }
        }
    }

    /**
     * An unique ID for this port
     *
     * @return
     */
    override fun toString(): String {
        return name
    }

    /**
     * Acquire lock on this instance of port handler with given controller
     * object. If port is currently locked by another controller, the wait until it is released.
     * Only the same controller  can release the port.
     *
     * @param controller
     * @throws hep.dataforge.exceptions.PortException
     */
    @Throws(PortException::class)
    fun holdBy(controller: PortController?) {
        assert(controller != null)
        if (!isOpen) {
            open()
        }

        execute {
            try {
                portLock.lockInterruptibly()
            } catch (ex: InterruptedException) {
                logger.error("Lock on port {} is broken", toString())
                throw RuntimeException(ex)
            }
        }
        logger.debug("Locked by {}", controller)
        this.controller = controller
    }


    /**
     * Receive a single byte
     */
    fun receive(byte: Byte) {
        controller?.accept(byte)
    }

    /**
     * Receive an array of bytes
     */
    fun receive(bytes: ByteArray) {
        controller?.accept(bytes)
    }

    /**
     * send the message to the port
     *
     * @param message
     * @throws hep.dataforge.exceptions.PortException
     */
    @Throws(PortException::class)
    protected abstract fun send(message: String)

    /**
     * Send the message if the controller is correct
     *
     * @param controller
     * @param message
     * @throws PortException
     */
    @Throws(PortException::class)
    fun send(controller: PortController?, message: String) {
        if (this.controller == null || controller === this.controller) {
            send(message)
        } else {
            throw PortException("Port locked by another controller")
        }
    }

    /**
     * Release hold of this portHandler from given controller.
     *
     * @param controller
     * @throws PortLockException in case given holder is not the one that holds
     * handler
     */
    @Synchronized
    @Throws(PortLockException::class)
    fun releaseBy(controller: PortController?) {
        if (isLocked) {
            assert(controller != null)
            if (controller == this.controller) {
                this.controller = null
                execute {
                    portLock.unlock()
                    logger.debug("Unlocked by {}", controller)
                }
            } else {
                throw PortLockException("Can't unlock port with wrong controller")
            }
        } else {
            logger.warn("Attempting to release unlocked port")
        }
    }


    @Throws(Exception::class)
    override fun close() {
        executor.shutdown()
    }

    class PortTimeoutException(private val timeout: Duration) : PortException() {
        override val message: String = String.format("The timeout time of '%s' is exceeded", timeout)
    }

    override fun toMeta(): Meta {
        return meta
    }

    companion object {
        /**
         * The definition of default phrase condition
         *
         * @return
         */
        private val DEFAULT_PHRASE_CONDITION = { str: String -> str.endsWith("\n") }
    }
}
